scripts/west_commands/runners/intel_adsp.py

```diff
@@ -138,8 +140,14 @@ def flash(self, **kwargs):
 
         self.logger.debug(f"rcmd: {self.run_cmd}")
 
+        # Launch the log monitor thread before fw is downloaded to avoid log missing
+        log_monitor_thread = threading.Thread(target=self.log_monitor)
+        log_monitor_thread.start()
+        time.sleep(0.5)
```

Comment:
I think this magic sleep is a bit hacky, and it's unclear why you need a separate thread just to ensure that the logging subprocess has already been started.

Why not:

- use `self.popen_ignore_int()` instead to start the logging process manually
- handle the run command as-is from the same thread using self.check_call() as is done now
- terminate the popen object afterwards

```py
        popen = self.popen_ignore_int(... log command arguments go here ...)

        try:
            self.check_call(self.run_cmd)
        finally:
            popen.terminate()
```

As a bonus, if your logging command has some sort of output it can emit to make it clear that's ready, you can poll for that before spawning `self.run_cmd`.

Reply:
The `self.check_call()` is a blocking call. Currently it is used to launch the `run_cmd` **first** and **then** `log_cmd`.  So I use a separated thread for `log_cmd` so that it can be started before `run_cmd` via the `check_call()`.

I didn't notice there's another non-blocking `popen_ignore_int()`. It does can be used like in you code snippet.
But maybe we still need some extra delay before the `run_cmd`  just to ensure that the log process is fully ready for log collection before the firmware download is carried out by `run_cmd`.

Reply:
@smrtos can you address whether this is possible from my previous comment please? I believe it is relevant to your latest response.

> As a bonus, if your logging command has some sort of output it can emit to make it clear that's ready, you can poll for that before spawning self.run_cmd.

I think we should always avoid magic delays whenever possible. Half a second is enough if the system isn't overloaded, but there is still a correctness problem here.

Reply:
I got your point. It will be better if the logging thread can explicitly indicate that it is ready. But the latter `run_cmd` is to flash the board. There will be no log output from the board until the `run_cmd` is finished. So I haven't found a way for the explicit indication. 
Now I turn to check the `is_alive()` of the logging thread. And because the flash operation will take some time to finish, I think we can ensure the logging thread will be ready this way.
Though there's still some sleep, it is not that magic now.

Reply:
This explanation belongs to the source code, otherwise most people reading the latter will think like @mbolivar-nordic

Reply:
> This explanation belongs to the source code, otherwise most people reading the latter will think like @mbolivar-nordic

Indeed. I added it to the comment. Thanks.

Reply:
@mbolivar-nordic  ping